lib/multicell_location/multicell_location.c

```diff
@@ -0,0 +1,230 @@
+{
+	int err, fd, bytes;
+	size_t offset = 0;
+	struct addrinfo *res;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM,
+	};
+
+	err = getaddrinfo(location_service_get_hostname(), NULL, &hints, &res);
+	if (err) {
+		LOG_ERR("getaddrinfo() failed, error: %d", err);
+		return err;
+	}
+
+	if (IS_ENABLED(CONFIG_MULTICELL_LOCATION_LOG_LEVEL_DBG)) {
+		char ip[INET6_ADDRSTRLEN];
+
+		inet_ntop(AF_INET, &((struct sockaddr_in *)res->ai_addr)->sin_addr,
+			  ip, sizeof(ip));
+		LOG_DBG("IP address: %s", log_strdup(ip));
+	}
+
+	((struct sockaddr_in *)res->ai_addr)->sin_port = htons(HTTPS_PORT);
+
+	fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TLS_1_2);
+	if (fd == -1) {
+		LOG_ERR("Failed to open socket, errno: %d", errno);
+		err = -errno;
+		goto clean_up;
+	}
+
+	/* Setup TLS socket options */
+	err = tls_setup(fd);
+	if (err) {
+		goto clean_up;
+	}
+
+	err = connect(fd, res->ai_addr, sizeof(struct sockaddr_in));
+	if (err) {
+		LOG_ERR("connect() failed, errno: %d", errno);
+		err = -errno;
+		goto clean_up;
+	}
+
+	do {
+		bytes = send(fd, &request[offset], request_len - offset, 0);
+
+		if (bytes < 0) {
+			LOG_ERR("send() failed, errno: %d", errno);
+			err = -errno;
+			goto clean_up;
+		}
+
+		offset += bytes;
+	} while (offset < request_len);
+
+	LOG_DBG("Sent %d bytes", offset);
+
+	bytes = recv(fd, recv_buf, sizeof(recv_buf), 0);
+	if (bytes < 0) {
+		LOG_ERR("recv() failed, err %d", errno);
+		err = -errno;
+		goto clean_up;
+	} else if (bytes == 0) {
+		LOG_ERR("No data received");
+		err = -ENODATA;
+		goto clean_up;
+
+	}
+
+	LOG_DBG("Received %d bytes", bytes);
+	LOG_DBG("HTTP response: \n%.*s\n", bytes, recv_buf);
+	LOG_DBG("Closing socket");
+
+	err = 0;
+
+clean_up:
+	freeaddrinfo(res);
+	(void)close(fd);
+
+	return err;
+}
+
+int multicell_location_get(const struct lte_lc_cells_info *cell_data,
+			   struct multicell_location *location)
+{
+	int err;
+	static bool cjson_is_init;
+
+	if ((cell_data == NULL) || (location == NULL)) {
+		return -EINVAL;
+	}
+
+	if (!cjson_is_init) {
+		cJSON_Init();
+
+		cjson_is_init = true;
+	}
+
+	err = location_service_generate_request(cell_data, http_request,
+						sizeof(http_request));
+	if (err) {
+		LOG_ERR("Failed to generate HTTP request, error: %d", err);
+		return err;
+	}
+
+	LOG_DBG("Generated request: \n%s", log_strdup(http_request));
+
+	err = execute_http_request(http_request, strlen(http_request));
+	if (err) {
+		LOG_ERR("Failed to send HTTP request");
+		return err;
+	}
+
+	err = location_service_parse_response(recv_buf, location);
+	if (err) {
+		LOG_ERR("Failed to parse HTTP response");
+		return -ENOMSG;
+	}
+
+	return 0;
+}
+
+int multicell_location_provision_certificate(bool overwrite)
+{
+	int err;
+	bool exists;
+	uint8_t unused;
+	const char *certificate = location_service_get_certificate();
+
+	if (certificate == NULL) {
+		LOG_ERR("No certificate was provided by the location service");
+		return -EFAULT;
+	}
+
```

Comment:
The API states that provisioning depends on LTE not being active. Since we are already depending on the link controller why no add a check here and return error if the link is in fact active?

Reply:
Added an assert to the sample to make sure auto connect is not used, can add a check here to later on